core/rapid_test.go

```diff
@@ -51,48 +53,157 @@ func (m *mockInsertedProposals) insertProposal(
+}
+
+// propertyTestSetup contains randomly-generated data for rapid testing
+type propertyTestSetup struct {
+	sync.Mutex
+
+	// nodes is the total number of nodes
+	nodes uint64
 
 	// desiredHeight is the desired height number
 	desiredHeight uint64
+
+	// events is the mapping between the current height and its rounds
+	events [][]propertyTestEvent
+
+	currentHeight map[int]uint64
+	currentRound  map[int]uint64
+}
+
+func (s *propertyTestSetup) setRound(nodeIndex int, round uint64) {
+	s.Lock()
+	s.currentRound[nodeIndex] = round
+	s.Unlock()
+}
+
+func (s *propertyTestSetup) incHeight() {
+	s.Lock()
+	for nodeIndex := 0; uint64(nodeIndex) < s.nodes; nodeIndex++ {
+		s.currentHeight[nodeIndex]++
+		s.currentRound[nodeIndex] = 0
+	}
+	s.Unlock()
+}
+
+func (s *propertyTestSetup) getEvent(nodeIndex int) propertyTestEvent {
+	s.Lock()
+
+	height := int(s.currentHeight[nodeIndex])
+	roundNumber := int(s.currentRound[nodeIndex])
+	var round propertyTestEvent
+	if roundNumber >= len(s.events[height]) {
+		round = s.events[height][len(s.events[height])-1]
+	} else {
+		round = s.events[height][roundNumber]
+	}
+
+	s.Unlock()
+
+	return round
+}
+
+func (s *propertyTestSetup) lastRound(height uint64) propertyTestEvent {
+	return s.events[height][len(s.events[height])-1]
+}
+
+func (s *propertyTestSetup) lastRoundQuorum(height uint64) int64 {
+	lastRound := s.lastRound(height)
+	return int64(s.nodes - lastRound.badNodes())
 }
 
 // generatePropertyTestEvent generates propertyTestEvent model
-func generatePropertyTestEvent(t *rapid.T) *propertyTestEvent {
+func generatePropertyTestEvent(t *rapid.T) *propertyTestSetup {
 	var (
-		numNodes             = rapid.Uint64Range(4, 15).Draw(t, "number of cluster nodes")
-		numByzantineNodes    = rapid.Uint64Range(0, maxFaulty(numNodes)).Draw(t, "number of byzantine nodes")
-		silentByzantineNodes = rapid.Uint64Range(0, numByzantineNodes).Draw(t, "number of silent byzantine nodes")
-		desiredHeight        = rapid.Uint64Range(10, 20).Draw(t, "minimum height to be reached")
+		numNodes      = rapid.Uint64Range(4, 15).Draw(t, "number of cluster nodes")
+		desiredHeight = rapid.Uint64Range(1, 5).Draw(t, "minimum height to be reached")
+		maxBadNodes   = maxFaulty(numNodes)
 	)
 
-	return &propertyTestEvent{
-		nodes:                numNodes,
-		byzantineNodes:       numByzantineNodes,
-		silentByzantineNodes: silentByzantineNodes,
-		badByzantineNodes:    numByzantineNodes - silentByzantineNodes,
-		desiredHeight:        desiredHeight,
+	setup := &propertyTestSetup{
+		nodes:         numNodes,
+		desiredHeight: desiredHeight,
+		events:        make([][]propertyTestEvent, desiredHeight),
+		currentHeight: map[int]uint64{},
+		currentRound:  map[int]uint64{},
+	}
+
+	for height := uint64(0); height < desiredHeight; height++ {
+		var round uint64
+		for {
+			numByzantineNodes := rapid.
+				Uint64Range(0, maxBadNodes).
+				Draw(t, fmt.Sprintf("number of byzantine nodes for height %d on round %d", height, round))
```

Comment:
I am not sure that it'll result in reproducible rapid test. You might need later to switch to generating slices.

Reply:
It generates random data every time so everything should be ok